### PR TITLE
Add support for Sensu etcd replicator

### PR DESCRIPTION
**Pull request by treydock**
# Pull Request Checklist

## Description

Add new type `sensu_etcd_replicator`.

## Related Issue

Part of #1174 

## Motivation and Context

Sensu Go added support for Etcd replicators and this adds a custom type to manage replicators.

### CODE REVIEW

File: lib/puppet/provider/sensu_etcd_replicator/sensuctl.rb [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    end
+    begin
+      sensuctl_delete('EtcdReplicator', resource[:name], nil, metadata, spec, 'federation/v1')
+    rescue Exception => e
+      raise Puppet::Error, "sensuctl delete check #{resource[:name]} failed\nError message: #{e.message}"
+    end
+    @property_hash.clear
+  end
+end
+
```
Comment (ghoneycutt) [inline lib/puppet/provider/sensu_etcd_replicator/sensuctl.rb]: trailing new line

**Comment by ghoneycutt** [comment]
Looks great, just need a few small things

- [ ] README entry
- [ ] `tests/` manifest
- [ ] vagrant

**Comment by treydock** [comment]
Added vagrant setup that creates two separate backends and creates a replicator on one and verifies the role created on a node is replicator to the other node.  Also updated README.

Added fix to API validator private type to query /version as ran into some odd bugs with /health and both are a way to tell if API is ready to accept requests.

File: README.md [hunk 2]
```diff
@@ -576,6 +585,40 @@ sensu::backend::config_hash:
...
+```puppet
+class { '::sensu::backend':
+  config_hash => {
+    'etcd-listen-client-urls' => ["http://localhost:2379","http://${facts['networking']['interfaces']['eth1']['ip']}:2379"],
+  },
+}
+```
+
+Next configure the Etcd replicator on the backend you wish to push resources from.
+In the following example all defined `Role` resources will be replicated to the backend at the IP address 192.168.52.30.
```
Comment (ghoneycutt) [inline README.md]: This is a bit confusing with roles and profiles. Not sure what Role means here.

File: README.md [hunk 3]
```diff
@@ -576,6 +585,40 @@ sensu::backend::config_hash:
...
 
+### Sensu backend federation
+
+Currently the federation support within this module involves configuring Etcd replicators. This allows resources to be sent from one Sensu cluster to another cluster.
+
+It's necessary that Etcd be listening on an interface that can be accessed by other Sensu backends.
+
+First configure backend Etcd to listen on an interface besides localhost:
+
+```puppet
```
Comment (ghoneycutt) [inline README.md]: do we have multiple roles for the sensu backends now such that one is where writes to etcd happen and the other nodes read changes or do all nodes support writes and then sync those between themselves?

File: Vagrantfile [hunk 4]
```diff
@@ -73,6 +73,31 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
...
+    backend.vm.box = "centos/7"
+    backend.vm.hostname = 'sensu-backend-peer1.example.com'
+    backend.vm.network :private_network, ip: '192.168.52.30'
+    backend.vm.network :forwarded_port, guest: 2380, host: 2383, auto_correct: true
+    backend.vm.network :forwarded_port, guest: 3000, host: 3003, auto_correct: true
+    backend.vm.network :forwarded_port, guest: 8080, host: 8084, auto_correct: true
+    backend.vm.network :forwarded_port, guest: 8081, host: 8086, auto_correct: true
+    backend.vm.provision :shell, :path => "tests/provision_basic_el.sh"
+    backend.vm.provision :shell, :inline => "puppet apply /vagrant/tests/sensu-backend-federated-cluster.pp"
+    backend.vm.provision :shell, :inline => "sensuctl role list"
```
Comment (ghoneycutt) [inline Vagrantfile]: a comment in here like if this works you should see X output at the end of the provisioning step would be helpful.

File: README.md (hunk 2 above)
Comment (treydock) [reply README.md]: `Role` is the Sensu Go resource, and copied from Sensu Go documentation.

File: README.md (hunk 3 above)
Comment (treydock) [reply README.md]: Not sure I follow the question. Based on Sensu Go docs it's designed so you can setup RBAC resources on one cluster and replicate those RBAC resources to another Sensu Go cluster (roles, role bindings, etc).

File: Vagrantfile (hunk 4 above)
Comment (treydock) [reply Vagrantfile]: I am documenting that in README with latest commit just pushed to support SSL

**Comment by treydock** [comment]
Updated the vagrant environment to use SSL for etcd which is the recommended method documented by Sensu Go. We can't use Puppet certs for Etcd because Etcd requires IP SAN and unable to generate certs with Puppet that set IP SANs. The Sensu Go docs document how to generate certs for Etcd and that was the approach I used and put into `tests/etcd-ssl`.

**Comment by treydock** [comment]
Running vagrant per README generates this upon second execution of provision:

```
==> sensu-backend-federated1: Running provisioner: shell...
    sensu-backend-federated1: Running: inline script
    sensu-backend-federated1:   Name   Namespace   Rules  
    sensu-backend-federated1:  ────── ─────────── ─────── 
    sensu-backend-federated1:   test   default         1  
```

That role was created on the other vagrant box and replicated to this box via etcd replicator.

File: README.md [hunk 5]
```diff
@@ -128,6 +129,18 @@ vagrant up sensu-backend-peer1 sensu-backend-peer2
...
+
+Multiple Vagrant boxes are available for testing a Sensu Go federated cluster.
+First build and provision both then provision the first a second time to view that the custom role was replicated.
+
+```base
+vagrant up sensu-backend-federated1 sensu-backend-federated2
+vagrant provision sensu-backend-federated1
+```
+
+The `provision` command should output from `sensuctl` the `test` Sensu Go Role that was created on the other backend.
```
Comment (ghoneycutt) [inline README.md]: Please provide sample output so users can confirm they are getting the desired effect.

File: README.md (hunk 3 above)
Comment (ghoneycutt) [reply README.md]: All resources or specific ones?

Curious how replication works and if there is a master in the cluster or if all nodes are equal.

File: tests/README.md [hunk 6]
```diff
@@ -24,6 +24,51 @@ Copy certs from vagrant instance to this repo
...
+sudo su -
+```
+
+Bootstrap SSL cert tool
+
+```
+yum install golang-bin
+cd /root
+git clone https://github.com/cloudflare/cfssl.git
+cfssl/
```
Comment (ghoneycutt) [inline tests/README.md]: i think you mean `cd cfssl`

File: README.md (hunk 3 above)
Comment (treydock) [reply README.md]: I don't know for sure, just docs reference RBAC resources but don't mention any limits. I think it's just sending resources from one backend to another when they are not in the same cluster.  I tried to make this PR just a tool to manage the replicator and leave the details and explanations of usage be covered in Sensu Go docs, so far all I've tested is sending RBAC resources per Sensu Go docs. The examples I'm providing and using are all from Sensu Go docs, didn't want to try and deviate too much from that.

File: tests/README.md (hunk 6 above)
Comment (treydock) [reply tests/README.md]: typo fixed

File: README.md (hunk 5 above)
Comment (treydock) [reply README.md]: Added output that should be printed.

**Pull request merged**